libraries/AP_CANManager/AP_CANManager.cpp

```diff
@@ -424,13 +435,102 @@ void AP_CANManager::handle_can_frame(const mavlink_message_t &msg) const
     hal.can[p.bus]->send(frame, AP_HAL::native_micros64() + timeout_us, AP_HAL::CANIface::IsMAVCAN);
 }
 
+/*
+  handle a CAN_FILTER_MODIFY packet
+ */
+void AP_CANManager::handle_can_filter_modify(const mavlink_message_t &msg)
+{
+    mavlink_can_filter_modify_t p;
+    mavlink_msg_can_filter_modify_decode(&msg, &p);
+    const int8_t bus = int8_t(p.bus)-1;
+    if (bus >= HAL_NUM_CAN_IFACES || hal.can[bus] == nullptr) {
+        return;
+    }
+    if (p.num_ids > ARRAY_SIZE(p.ids)) {
+        return;
+    }
+    uint16_t *new_ids = nullptr;
+    uint16_t num_new_ids = 0;
+    WITH_SEMAPHORE(can_forward.sem);
+
+    switch (p.operation) {
+    case CAN_FILTER_REPLACE: {
+        if (p.num_ids == 0) {
+            can_forward.num_filter_ids = 0;
+            delete[] can_forward.filter_ids;
+            can_forward.filter_ids = nullptr;
+            return;
+        }
+        new_ids = new uint16_t[p.num_ids];
+        if (new_ids != nullptr) {
+            num_new_ids = p.num_ids;
+            memcpy((void*)new_ids, (const void *)p.ids, p.num_ids*sizeof(uint16_t));
+            // sort the list, so we can bisection search
+            insertion_sort_uint16(new_ids, num_new_ids);
+        }
+        break;
+    }
+    case CAN_FILTER_ADD: {
+        new_ids = new uint16_t[can_forward.num_filter_ids+p.num_ids];
+        if (new_ids == nullptr) {
+            return;
+        }
+        memcpy(new_ids, can_forward.filter_ids, can_forward.num_filter_ids*sizeof(uint16_t));
```

Comment:
that just wastes flash space (zero length copter to any address is fine)

Reply:
zero length copy is fine as long as the pointers are valid. Highly unlikely that it will actually fail, but It seems to be declared as an undefined behaviour per C standard if the pointers are invalid as well.

Reply:
i think we rely on it quite a bit in ardupilot, but I've added in the check anyway. I think it is wasteful, but not by much